Lib9c/Action/ItemEnhancement.cs

```diff
@@ -211,62 +221,91 @@ public override IAccountStateDelta Execute(IActionContext context)
                 );
             }
 
             if (materialEquipment.level != enhancementEquipment.level)
             {
-                // Invalid level
                 throw new InvalidMaterialException(
-                    $"{addressesHex}Aborted as levels of the equipment to enhance ({enhancementEquipment.level}) and a material ({materialEquipment.level}) does not match."
+                    $"{addressesHex}Aborted as levels of the equipment to enhance ({enhancementEquipment.level})" +
+                    $" and a material ({materialEquipment.level}) does not match."
                 );
             }
             sw.Stop();
             Log.Verbose("{AddressesHex}ItemEnhancement Get Material: {Elapsed}", addressesHex, sw.Elapsed);
+
             sw.Restart();
-            materialEquipment.Unequip();
+            // Subtract required action point
+            var requiredActionPoint = GetRequiredAp();
+            if (avatarState.actionPoint < requiredActionPoint)
+            {
+                throw new NotEnoughActionPointException(
+                    $"{addressesHex}Aborted due to insufficient action point: {avatarState.actionPoint} < {requiredActionPoint}"
+                );
+            }
+            avatarState.actionPoint -= requiredActionPoint;
 
-            enhancementEquipment.Unequip();
+            // TransferAsset (NCG)
+            var requiredNcg = row.Cost;
+            if (requiredNcg > 0)
+            {
+                states = states.TransferAsset(ctx.Signer, BlacksmithAddress, states.GetGoldCurrency() * requiredNcg);
+            }
 
-            enhancementEquipment = ItemEnhancement.UpgradeEquipment(enhancementEquipment);
+            // Unequip items
+            materialEquipment.Unequip();
+            enhancementEquipment.Unequip();
 
-            var requiredBlockIndex = ctx.BlockIndex + RequiredBlockCount;
+            // Equipment level up & Update
+            var equipmentResult = GetEnhancementResult(row, ctx.Random);
+            if (equipmentResult != EnhancementResult.Fail)
+            {
+                enhancementEquipment.LevelUpV2(ctx.Random, row, equipmentResult == EnhancementResult.GreatSuccess);
+            }
+            var requiredBlockCount = GetRequiredBlockCount(row, equipmentResult);
+            var requiredBlockIndex = ctx.BlockIndex + requiredBlockCount + GameConfig.RequiredAppraiseBlock;
             enhancementEquipment.Update(requiredBlockIndex);
+
+            // Remove material
+            avatarState.inventory.RemoveNonFungibleItem(materialId);
             sw.Stop();
             Log.Verbose("{AddressesHex}ItemEnhancement Upgrade Equipment: {Elapsed}", addressesHex, sw.Elapsed);
-            sw.Restart();
 
-            result.gold = requiredNCG;
+            // Send scheduled mail
+            var result = new ResultModel
+            {
+                preItemUsable = new Equipment((Dictionary)enhancementEquipment.Serialize()),
```

Comment:
i think preItemUsable is duplicated itemUsable(already enhancemented). i think Clone before LevelUpV2 call.